### ISSUE TITLE
Tips optimization for covariation

### ISSUE THREAD

**Issue by LeaFrock** [dotnet/roslyn#48432]
**Brief description:**

Recently I met a confusing problem about covariation. Details see the link below please,
[dotnet/csharplang/discussions/3985](https://github.com/dotnet/csharplang/discussions/3985)

**Languages applicable:**

C# (I don't know whether VB has covariation...)

**Code example that the analyzer should report:**

```C#
    public class SingleHashFunctionList<T> : IHashFunctionList  
        where T : IHashFunction //Notice this 'where'
    {
        private readonly IReadOnlyList<T> _hashFunctions;

        public SingleHashFunctionList()
        {
            // Skip init work of '_hashFunctions'
            HashFunctions = _hashFunctions; //Compiler error: ...cannot implictly convert...
        }

        public IReadOnlyList<IHashFunction> HashFunctions { get; }
    }
```
Currently here the compiler reports the error as `...cannot implictly convert...from...to...` and doesn't provide any infomation or solutions related to covariation.

I think smarter and more friendly tips should be like `Covariant type parameters must work with class. T might not be a class` And if possible, the analyzer can provide automatic solution which converts the `where T : IHashFunction` to `where T : class, IHashFunction` .

**Comment by huoyaoyuan** [comment]
This should be diagnostic clarity. The binding should know the failure reason is missing reference type constraint.
